src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs

```diff
@@ -213,6 +213,33 @@ public ImmutableArray<T> MoveToImmutable()
                 return new ImmutableArray<T>(temp);
             }
 
+            /// <summary>
+            /// Returns the current contents as an <see cref="ImmutableArray{T}"/> and sets the collection to a zero length array.
+            /// </summary>
+            /// <remarks>
+            /// If <see cref="Capacity"/> equals <see cref="Count"/>, the internal array will be extracted
+            /// as an <see cref="ImmutableArray{T}"/> without copying the contents. Otherwise, the contents
+            /// will be copied into a new array. The collection will then be set to a zero length array.
+            /// </remarks>
+            /// <returns>An immutable array.</returns>
+            public ImmutableArray<T> DrainToImmutable()
+            {
+                ImmutableArray<T> result;
+
+                if (Capacity == Count)
+                {
+                    result = MoveToImmutable();
+                }
+                else
+                {
+                    result = ToImmutable();
+                    _elements = ImmutableArray<T>.Empty.array!;
+                    _count = 0;
+                }
+
+                return result;
```

Comment:
```suggestion
                ImmutableArray<T> result = Capacity == Count ?
                    _elements :
                    ToArray();
                
                _elements = ImmutableArray<T>.Empty.array!;
                _count = 0;

                return result;
```
?

Reply:
Thanks for this coaching, @stephentoub! Refactored per suggestion.